README.md
# my_ddns

Open Source DDNS, with connectors

## Requirements

- Python 3^
- pip

## Installation

### General use case

The program use python 3, and the dependencies have to be installed manually in custom environments
Use pip and the requirements.txt file

## Usage

For now, connector used is hardcoded. Modify to meet your needs.

(Old mail in a box connector command, to be modified in short future :)
`python3 main.py {ddns domain} {miab domain} {username} {password}`

For a normal usage, the parameters have to be specified as `--parameter-name=parameter-value`

Here a PowerDNS command exemple:
`python3 main.py --domain="test.exemple.org." --instance-url="https://pdns.exemple.org" --zone="exemple.org" --user="user" --api-key="some_api_key"`

The script will automatically check each 300s (5min) for ip update

## Compatibility

You will find [here](compatibility.md) the basic compatibility list, with a little part of roadmap

CHANGELOG.md
# This file keep track of project changes

## UNRELEASED

### ADDED
- Cloudflare connector #11

### CHANGED
- Match connector type

### FIXED
- Check required connector's parameters

# [2.1.1] - 2023-12-18
### CHANGED
- Make dependabot respect git flow

# [2.1.0] - 2023-12-18
### ADDED
- Add basic dependabot conf

# [2.0.1] - 2023-12-18
### ADDED
- Add changelog file 

# [2.0.0] - 2022-09-20
### ADDED
- Read configuration from file

# [1.8.0] - 202-08-18
### ADDED
- add the ability to only update ipv4 or ipv6 #19

# [1.7.0] - 2022-08-04
### FIXED
- Fix MIAB connector #17

# [1.6.0] - 2022-08-03
### CHANGED
- when first dns update, do not update again without ip change


# [1.5.1] - 2022-08-03
### ADDED
- Add licence

# [1.5.0] - 2022-08-03
### ADDED
- Wiki: guide on writing connector #9

### FIXED
- Core: the dns update is always updated, this is now corrected

# [1.4.0] - 2022-08-03
### CHANGED
- The connectors should now only use parameters set in set_optional_parameter

# [1.3.0] - 2022-08-03
### ADDED
- Add support of ipv6
- Add PowerDNS connector #6

# [1.1.0] - 2022-02-15
### ADDED
- Update multiples domains for same ip at once

# [1.0.0] - 2022-01-02
### ADDED
- Mail in a box connector


main.py
import os.path
import time
import sys

from typing import List

from connector.MailInABoxConnector import MailInABoxConnector
from connector.PowerDnsConnector import PowerDnsConnector
from util.text import get_public_ip, get_public_ipv6, info


def parse_params(param_list: List[str]) -> dict:
    parsed_params_list = {}

    for p in param_list:
        if not p.__contains__('='):
            continue
        if p.startswith('--'):
            p = p.removeprefix('--')

        param: List[str] = p.split('=')
        parsed_params_list[str(param[0])] = str(param[1])

    return parsed_params_list


def main():
    current_ip = ""
    current_ipv6 = ""

    parameters = {}

    if len(sys.argv) > 1:
        parameters = parse_params(sys.argv)
    else:
        # Parameters are stored in a config file
        params = []
        with open('./config.txt') as file:
            for line in file:
                params.append(line.rstrip())

        parameters = parse_params(params)

    # Set connector independant vars
    domain = str(parameters.get('domain'))
    file = str(parameters.get('file'))

    # Check if domain parameter is a filename
    if domain is None and file is not None:
        if os.path.exists('./' + file):
            domain = []

            with open('./' + file) as file:
                domain.append(line.rstrip() for line in file)

    connector = PowerDnsConnector()

    required_parameters = connector.get_optionals_parameters()
    for parameter in required_parameters:
        if parameters.get(parameter) is None:
            raise f"Required parameter {parameter} is missing"

    # For each parameter, set it in the connector
    for param in parameters:
        connector.set_optional_parameter(param, parameters.get(param))

    # Tableau de la liste des ip à mettre à jour
    ip_to_update_list: list = []

    should_update_dns = False
    
    should_update_ipv4 = True
    should_update_ipv6 = True

    only_update = parameters.get('only-update')
    if only_update is not None:
        #  --only-update=ipv4 or ipv6
        if only_update == 'ipv4':
            should_update_ipv6 = False
        else:
            should_update_ipv4 = False

    while True:
        last_ip = get_public_ip()
        last_ipv6 = get_public_ipv6()

        if last_ip != current_ip and should_update_ipv4:
            ip_to_update_list.append(last_ip)
            current_ip = last_ip
            should_update_dns = True
            info(f"Public ip is no longer anymore {current_ip}, now it's {last_ip}")

        if last_ipv6 != current_ipv6 and should_update_ipv6:
            ip_to_update_list.append(last_ipv6)
            current_ipv6 = last_ipv6
            should_update_dns = True
            info(f"Public ipv6 is no longer anymore {current_ipv6}, now it's {last_ipv6}")

        if should_update_dns:
            if isinstance(domain, list):
                for d in domain:
                    for ip in ip_to_update_list:
                        connector.update_dns(d, ip)
            else:
                for ip in ip_to_update_list:
                    connector.update_dns(domain, ip)

            should_update_dns = False

        time.sleep(300)


if __name__ == "__main__":
    main()


util/text.py
import ipaddress

import requests


def get_public_ip():
    r = requests.get(url="https://api.ipify.org/?format=json")
    data = r.json()

    return data['ip']


def get_public_ipv6():
    r = requests.get(url="https://api64.ipify.org/?format=json")
    data = r.json()
    ip = ''

    try:
        ip = ipaddress.IPv6Address(data['ip'])
    except ValueError:
        pass

    return ip


class BGColors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKCYAN = '\033[96m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'


def info(text: str):
    print(f"{BGColors.OKBLUE}{text}{BGColors.ENDC}")


def warn(text: str):
    print(f"{BGColors.WARNING}{text}{BGColors.ENDC}")


def error(text: str):
    print(f"{BGColors.FAIL}{text}{BGColors.ENDC}")


def success(text: str):
    print(f"{BGColors.OKGREEN}{text}{BGColors.ENDC}")


util/dns/Record.py
import ipaddress
from enum import Enum
from typing import Union


class RecordType(Enum):
    """
    Enum to use when define a dns record type
    """
    A = 1
    AAAA = 2
    PTR = 3
    CNAME = 4
    NS = 5

    # Keep last
    DEFAULT = 0


class Record:
    """
    Class which represent a DNS record
    """

    def __init__(self, content: str, record_type: RecordType = RecordType.DEFAULT, ttl: int = 86400):
        """
        Initialize a record with the given parameters.
        By default, the record type is not valid, as it has to be verified by the programmer.

        """
        self.record_type = record_type
        self.ttl = ttl
        self.content = content

        self.record_type = self.get_record_type()

    def is_ip(self) -> bool:
        """
        Check if the content correspond to a valid ipv4 or ipv6 address
        """
        valid = False

        try:
            ipaddress.ip_address(self.content)
            valid = True
        except ValueError:
            pass

        return valid

    def get_ip(self) -> Union[ipaddress.IPv4Address, ipaddress.IPv6Address]:
        return ipaddress.ip_address(self.content)

    def get_record_type(self) -> RecordType:
        """
        Return the record type corresponding to value passed in constructor (A, AAAA or CNAME)

        TODO: check for other types (NS,PTR, ...)
        """
        try:
            ipaddress.IPv4Address(self.content)
            return RecordType.A
        except ValueError:
            try:
                ipaddress.IPv6Address(self.content)
                return RecordType.AAAA
            except ValueError:
                return RecordType.CNAME


connector/CloudflareConnector.py
import json

import requests

from connector.AbstractConnector import AbstractConnector
from util.dns.Record import Record
from util.text import success, error


class CloudflareConnector(AbstractConnector):
    """
    Cloudflare connector
    @author Romain NEIL
    """

    def __init__(self):
        self.instance_url = "api.cloudflare.com/client"
        self.api_version = "v4"

        self.params = {}

    def get_optionals_parameters(self) -> list[str]:
        return [
            'has_api_key',
            'auth_param'
            'domain',
            'ttl',
            'proxied'
        ]

    def set_optional_parameter(self, param_name: str, param_value: str):
        self.params[param_name] = param_value

    def build_headers(self):
        auth_header = 'X-Auth-Email'

        if self.params.get('has_api_key', False):
            auth_header = 'X-Api-Key'

        return {
            auth_header: self.params.get('auth_param')
        }

    def build_params(self, domain: str, record: Record) -> str:
        return json.dumps({
            "content": record.get_ip(),
            "name": domain,
            "type": record.get_record_type().name,
            "ttl": self.params.get('ttl', 3600),
            "proxied": self.params.get('proxied', 'false')
        })

    def update_dns(self, domain: str, ip: str):
        record = Record(str(ip))

        r = requests.put(
            url=f"https://{self.instance_url}/{self.api_version}/zones/zone_identifier/dns_records/identifier",
            headers=self.build_headers()
        )

        if 200 <= r.status_code < 300:
            success(f"[SUCCESS]: Update domain to {record.content}")
        else:
            resp = json.loads(r.content)

            error(f"[ERROR]: Cannot update domain")
            error(f"{resp.error.message}")


connector/MailInABoxConnector.py
import requests

from connector.AbstractConnector import AbstractConnector
from util.text import success, error


class MailInABoxConnector(AbstractConnector):
    """
    Connector for a mail in a box server
    """

    def __init__(self):
        self.instance_url = None
        self.password = None
        self.username = None

        self.params = {}

    def get_optionals_parameters(self) -> list[str]:
        return [
            'username',
            'password',
            'instance-url'
        ]

    def set_optional_parameter(self, param_name: str, param_value: str):
        self.params[param_name] = param_value

    def update_dns(self, domain, ip):
        r = requests.put(
            url=f"https://{self.params.get('password')}/admin/dns/custom/{domain}",
            auth=(self.params.get('username'), self.params.get('password'))
        )

        if 200 <= r.status_code < 300:
            success(f"[SUCCESS] Make new ip point to {domain}")
        else:
            error(f"Error: cannot update url. Reason : {r.reason}")


connector/PowerDnsConnector.py
import datetime
import json

import requests

from connector.AbstractConnector import AbstractConnector
from util.dns.Record import Record
from util.text import success, error, warn, info


class PowerDnsConnector(AbstractConnector):
    """
    Power DNS Connector
    @author Romain NEIL
    """

    def __init__(self):
        self.instance_url = None
        self.password = None
        self.username = None

        self.params = {}

    def get_optionals_parameters(self):
        return [
            'api-key',
            'domain',
            'zone'
        ]

    def set_optional_parameter(self, param_name: str, param_value: str):
        self.params[param_name] = param_value

    def build_params(self, domain: str, record: Record) -> str:
        return json.dumps({
            "rrsets":
                [{
                    "name": domain,
                    "type": str.upper(record.record_type.name),
                    "ttl": record.ttl,
                    "changetype": "REPLACE",
                    "records": [{
                        "content": record.content,
                        "disabled": False
                    }],
                    "comments": [{
                        "content": "modify record",
                        "account": f"{self.params.get('user')}",
                        "modified_at": int(round(datetime.datetime.now().timestamp()))
                    }]
                }]
        })

    def update_dns(self, domain, ip):
        record = Record(str(ip))

        data = self.build_params(domain, record)

        r = requests.patch(
            url=f"{self.params.get('instance-url')}/api/v1/servers/localhost/zones/{self.params.get('zone')}",
            headers={"X-API-Key": self.params.get('api-key')},
            data=data
        )

        if 200 <= r.status_code < 300:
            success(f"[SUCCESS] Make domain point to {record.content}")
        else:
            error(f"[ERROR] cannot update url. Reason : {r.reason}")
            warn(f"The url was {r.url}")
            info(f"The json object was : {data}")
